contracts/Compass.vy

```diff
@@ -104,39 +140,63 @@ def __init__(_compass_id: bytes32, _event_id: uint256, _gravity_nonce:uint256, v
     self.last_gravity_nonce = _gravity_nonce
     log ValsetUpdated(new_checkpoint, valset.valset_id, _event_id)
 
-# utility function to verify EIP712 signature
-@internal
-@pure
-def verify_signature(signer: address, hash: bytes32, sig: Signature) -> bool:
-    message_digest: bytes32 = keccak256(concat(convert("\x19Ethereum Signed Message:\n32", Bytes[28]), hash))
-    return signer == ecrecover(message_digest, sig.v, sig.r, sig.s)
-
-# consensus: validator set and signatures
-# hash: what we are checking they have signed
-@internal
-def check_validator_signatures(consensus: Consensus, hash: bytes32):
-    i: uint256 = 0
-    cumulative_power: uint256 = 0
-    for sig in consensus.signatures:
-        if sig.v != 0:
-            assert self.verify_signature(consensus.valset.validators[i], hash, sig), "Invalid Signature"
-            cumulative_power += consensus.valset.powers[i]
-            if cumulative_power >= POWER_THRESHOLD:
-                break
-        i = unsafe_add(i, 1)
-    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+# sends the current internal book keeping state to a new version of compass, along with all funds
+# left on the contract.
+# consensus: current validator set and signatures
+# new_compass: address of new compass to which to send the funds
+# message_id: unused message ID
+# deadline: message deadline
+# authority: intended message sender address
+# gas_estimate: gas estimation in wei
+@external
+@nonreentrant('lock')
+def send_books(consensus: Consensus, new_compass: address, message_id: uint256,deadline: uint256, authority: address, gas_estimate: uint256):
```

Comment:
Does it make sense to protect this function from being called twice (with different message_id)? If it somehow were to happen, it would clear the books on the new compass, right?

Reply:
Yeah, I'm torn here. I was considering adding a boolean flag that would just flip once this function was called. But we've seen transactions going through with the call down the chain failing. So I'm worried what happens in that case. We'll loose access to all funds. 

So, I opted for replayability instead. But maybe it's a better idea to prevent the `receive_books` function from being called more than once? I think that could work.